### ISSUE TITLE
Misses errors when inserting from list

### ISSUE THREAD

**Issue by fredcallaway** [JukkaL/mypy#945]
I think the following should raise an error, but it doesn't. I imagine it has to do with the call to `super`. Is there something obvious I'm missing?

```python
class IntList(list):
    def __init__(self, lst:List[int]) -> None:
        super(InheritanceList, self).__init__(lst)

int_list = IntList(list(range(10)))
int_list.append('f')  # no error here!
```

**Comment by jhance** [comment]
Don't think this should be an error. From the perspective of `mypy`, `IntList` is actually basically a `List[Any]`. You should be inheriting from `List[Int]` if you want it to know.

The thing thats *actually* suspect here is that `__init__` is accepting a `List[int]` where it ought to expect a `List[Any]` despite `List` not being covariant.

**Comment by fredcallaway** [comment]
Thanks jhance! I didn't realize that you could inherit from `List[int]`. Maybe it is due to my lack of understanding, but it wasn't clear to me from the docs that you could use the objects in `typing` that way.

I don't understand the point about `list` though. It looks like list can in fact take a `List[int]`. For example, this type checks:

```python
class InheritanceList(list):
    def __init__(self, lst:List[int]) -> None:
        super(InheritanceList, self).__init__(lst)

list_of_ints = [i for i in range(10)]  # type: List[int]
inlist = InheritanceList(list_of_ints)
inlist.append('f')
```

Secondly, I don't understand why a `List[int]` is not a `List[Any]` but an `Iterable[int]` is an `Iterable[Any]`. Why the lack of parallelism here? Is there documentation I can read other than the mypy docs? It seems like you understand the mypy type system much better than me, although I read all the documentation I could find.

Finally, would it be possible to create a class that inherits from `list` and type checks like `list`? That is, it would require all its elements to be of some arbitrary type? I'm imagining something like parameterized types in OCaml: e.g. `('a list)` Here's an example of the behavior I want:

```python

class InheritanceList(List):
    def __init__(self, lst) -> None:
        super(InheritanceList, self).__init__(lst)

list_of_ints = [i for i in range(10)]  # type: List[int]
inlist = InheritanceList(list_of_ints)
inlist.append('f')  # Error incompatible type "str"; expected "int"
```

This would parallel the current list behavior:
```python
normal_list = [1,2,3]
normal_list.append('a')  # Error incompatible type "str"; expected "int"
```

Thanks for your help!

**Comment by jhance** [comment]
> I don't understand the point about list though. It looks like list can in fact take a List[int].

It can, but only because the `list` constructor takes a `Sequence`.

> Secondly, I don't understand why a List[int] is not a List[Any] but an Iterable[int] is an Iterable[Any]. Why the lack of parallelism here? 

Mm I think this may only be documented in PEP484, but the difference is that `List` is not covariant (mutability causes issues here) but `Iterable` is.

> Finally, would it be possible to create a class that inherits from list and type checks like list? That is, it would require all its elements to be of some arbitrary type? I'm imagining something like parameterized types in OCaml: e.g. ('a list) Here's an example of the behavior I want:

Untested, but basically what you want is to have a type variable involved so that you don't implicitly get `Any`:

```
T = TypeVar('T')
class InheritanceList(List[T], Generic[T]):
    def __init__(self, lst: Iterable[T]) -> None:
        super(InheritanceList, self).__init__(list)
```

**Comment by JukkaL** [comment]
`Foo[Any]` is always compatible with `Foo[int]` (and vice versa), due to the way gradual typing works in mypy. In a strict mode we could potentially warn about this, but this shouldn't be the default behavior.

@jhance's last example is good and shows how you can define a generic type that behaves like `List[...]` during type checking. However, it should be possible to write it like this but this is not supported yet (#302):

```
T = TypeVar('T')

class InheritanceList(List[T]):
    def __init__(self, lst: Iterable[T]) -> None:
        super().__init__(lst)
```

**Comment by JukkaL** [comment]
Closing this as not a bug, but we could certainly improve the documentation and have a strict mode that would be more helpful in cases like these and disallow bare `list` as a base class.
